modules/dag/src/main/scala/org/tesselation/dag/l1/DAGBlockValidator.scala

```diff
@@ -0,0 +1,170 @@
+package org.tesselation.dag.l1
+
+import cats.effect.Async
+import cats.syntax.applicativeError._
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+import cats.syntax.semigroup._
+import cats.syntax.traverse._
+
+import scala.annotation.tailrec
+
+import org.tesselation.dag.l1.storage.{BlockStorage, TransactionStorage}
+import org.tesselation.domain.cluster.storage._
+import org.tesselation.infrastructure.db.schema.StoredAddress
+import org.tesselation.kryo.KryoSerializer
+import org.tesselation.schema.balance.Balance
+import org.tesselation.schema.transaction.{Transaction, TransactionReference}
+import org.tesselation.schema.{BlockValidator, nonNegBigIntSemigroup}
+import org.tesselation.security.signature.Signed
+import org.tesselation.security.{Hashed, SecurityProvider}
+
+import eu.timepit.refined.auto.autoInfer
+import eu.timepit.refined.numeric.NonNegative
+import eu.timepit.refined.refineV
+import io.estatico.newtype.ops.toCoercibleIdOps
+
+class DAGBlockValidator[F[_]: Async: KryoSerializer: SecurityProvider](
+  transactionStorage: TransactionStorage[F],
+  blockStorage: BlockStorage[F],
+  addressStorage: AddressStorage[F],
+  transactionValidator: TransactionValidator[F]
+) extends BlockValidator[F, DAGBlock] {
+
+  override def accept(signedBlock: Signed[DAGBlock]): F[Unit] =
+    for {
+      _ <- validateBlockSignatures(signedBlock).ifM(
+        Async[F].unit,
+        Async[F].raiseError[Unit](new Throwable("Invalid block signatures!"))
+      )
+      _ <- validateBlock(signedBlock.value).ifM(
+        Async[F].unit,
+        Async[F].raiseError[Unit](new Throwable("Invalid block data!"))
+      )
+      hashedBlock <- signedBlock.hashWithSignatureCheck.flatMap(_.liftTo[F])
+
+      hashedTransactions <- signedBlock.value.transactions.traverse(_.hashWithSignatureCheck.flatMap(_.liftTo[F]))
+
+      _ <- acceptTransactions(hashedTransactions)
```

Comment:
The transactions and block are being accepted separately here, what happens if there's an error in accepting the block? How do we roll the transactions back as a single operation? All the operations of validation before this are fine, but we need to explicitly capture errors associated with a single transact operation here, since this will eventually reflect database updates and should be treated that way. 

If there's a failure in updating anything, it should all be easy to roll back atomically. Not sure if that's within the scope of this PR.

Reply:
Yep I also thought about this problem during the implementing, so it's nice that you are also pointing it out. Didn't come up with a good solution though at that time. To be honest if the transactions acceptance passes (database is updated) and the block acceptance fails (which is a simple blocking Ref modification, so I think the chances are very low) I can't guarantee that any rollback operation on the database will succeed (or can I somehow?), so we may still end up in a broken state. I don't think we can atomically update database and blockStorage at the same time - or maybe we can but I don't know how. The only solution I can think of is balances hash as a part of snapshot and majority being picked. Now that we plan to switch to transactions for rewards this actually seems possible to achieve. Do you have some idea how that could be tackled?